### PR TITLE
[CAPI]Unicode support

### PR DESCRIPTION
**Pull request by riverlijunjie**
### Details:
 - Support unicode code for input path

### Tickets:
 - 89488

### CODE REVIEW

**Comment by riverlijunjie** [comment]
I also notice below C interfaces also will face unicode problem, but C++ doesn't provide explicitly unicode interface:

```
OPENVINO_C_API(ov_status_e) ov_core_create_with_config(const char* xml_config_file, ov_core_t** core);
OPENVINO_C_API(ov_status_e)  ov_core_compile_model_from_file(const ov_core_t* core,
                                const char* model_path,
                                const char* device_name,
                                const ov_properties_t* property,
                                ov_compiled_model_t** compiled_model);
```

Do we need support in C interface?

**Comment by ilyachur** [comment]
> I also notice below C interfaces also will face unicode problem, but C++ doesn't provide explicitly unicode interface:
> 
> ```
> OPENVINO_C_API(ov_status_e) ov_core_create_with_config(const char* xml_config_file, ov_core_t** core);
> OPENVINO_C_API(ov_status_e)  ov_core_compile_model_from_file(const ov_core_t* core,
>                                 const char* model_path,
>                                 const char* device_name,
>                                 const ov_properties_t* property,
>                                 ov_compiled_model_t** compiled_model);
> ```
> 
> Do we need support in C interface?

Yes I think we need also have these C API. Because user can skip `read_model` and put the model path to `compile_model` function

**Review by ilyachur — COMMENTED** [review]
Can we add tests on this API?

**Comment by riverlijunjie** [comment]
> > I also notice below C interfaces also will face unicode problem, but C++ doesn't provide explicitly unicode interface:
> > ```
> > OPENVINO_C_API(ov_status_e) ov_core_create_with_config(const char* xml_config_file, ov_core_t** core);
> > OPENVINO_C_API(ov_status_e)  ov_core_compile_model_from_file(const ov_core_t* core,
> >                                 const char* model_path,
> >                                 const char* device_name,
> >                                 const ov_properties_t* property,
> >                                 ov_compiled_model_t** compiled_model);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do we need support in C interface?
> 
> Yes I think we need also have these C API. Because user can skip `read_model` and put the model path to `compile_model` function

There are 2 ways to support it, which is better from your perspective?
1. direct unicode support in C libs
2. First enabled unicode in these C++ interfaces and then call them in C libs

**Comment by riverlijunjie** [comment]
> Can we add tests on this API?

yes, we need unicode test in CI, not only C++/python test, but also C interface test.
We can provide such test sample, but need more support from CI team to create unicode path.

**Comment by riverlijunjie** [comment]
Blocked by https://github.com/openvinotoolkit/openvino/pull/12887, in which some Windows build error can be avoided.

File: src/bindings/c/src/ov_core.cpp [hunk 1]
```diff
@@ -318,3 +329,90 @@ void ov_core_versions_free(ov_core_version_list_t* versions) {
         delete[] versions->versions;
     versions->versions = nullptr;
 }
+
+#ifdef OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
+inline std::string wstring_to_string(const std::wstring& wstr) {
```
Comment (ilyachur) [inline src/bindings/c/src/ov_core.cpp:334]: I think we shouldn't duplicate this function in C API.

You can link C API with `openvino::util` in order to use existed function.

File: src/bindings/c/src/ov_tensor.cpp [hunk 2]
```diff
@@ -8,7 +8,7 @@
 const std::map<ov_element_type_e, ov::element::Type> element_type_map = {
     {ov_element_type_e::UNDEFINED, ov::element::undefined},
     {ov_element_type_e::DYNAMIC, ov::element::dynamic},
-    {ov_element_type_e::BOOLEAN, ov::element::boolean},
+    {ov_element_type_e::BL1, ov::element::boolean},
```
Comment (ilyachur) [inline src/bindings/c/src/ov_tensor.cpp:11]: Why?

File: src/bindings/c/tests/ov_test.cpp [hunk 3]
```diff
@@ -24,7 +24,7 @@ std::string plugins_xml_std = TestDataHelpers::generate_ieclass_xml_path("plugin
 #endif
 const char* plugins_xml = plugins_xml_std.c_str();
 
-std::map<ov_element_type_e, size_t> element_type_size_map = {{ov_element_type_e::BOOLEAN, 8},
+std::map<ov_element_type_e, size_t> element_type_size_map = {{ov_element_type_e::BL1, 1},
```
Comment (ilyachur) [inline src/bindings/c/tests/ov_test.cpp:27]: Why did you rename element type?

File: src/bindings/c/tests/ov_test.hpp [hunk 4]
```diff
@@ -51,3 +75,84 @@ inline static std::vector<uint8_t> content_from_file(const char* filename, bool
     }
     return result;
 }
+
+#ifdef OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
+extern const std::vector<std::wstring> test_unicode_postfix_vector;
+inline std::string wstring_to_string(const std::wstring& wstr) {
```
Comment (ilyachur) [inline src/bindings/c/tests/ov_test.hpp:81]: We shouldn't duplicate existed functions

File: src/bindings/c/tests/ov_test.cpp (hunk 3 above)
Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.cpp:27]: In windows it reports below error:
  include\10.0.18362.0\um\winnt.h(1131): error C2365: 'BOOLEAN': redefinition; previous definition was 'enumerator'
  openvino\src\bindings\c\include\openvino/c/ov_common.h(105): note: see declaration of 'BOOLEAN'

File: src/bindings/c/tests/ov_test.hpp (hunk 4 above)
Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.hpp:81]: Maybe you are right, I just want to make the C test code more clear without too much dependent code.

Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.hpp]: done

File: src/bindings/c/src/ov_core.cpp (hunk 1 above)
Comment (riverlijunjie) [reply src/bindings/c/src/ov_core.cpp]: done

File: src/bindings/c/tests/ov_test.cpp (hunk 3 above)
Comment (ilyachur) [reply src/bindings/c/tests/ov_test.cpp:27]: Honestly I don't like that we need to rename data type. Do we have a stacktrace in order to find the place which causes the issue?

Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.cpp:27]: I agree with you, I also don't like such rename behavior. But in C scope, there is no namespace to avoid such conflict with windows header files. Let me check whether there is other way to avoid renaming.

Comment (ilyachur) [reply src/bindings/c/tests/ov_test.cpp:27]: I think also we need to understand the place which causes the issue

Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.cpp:27]: ok, I am setup local windows build machine to check it.

Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.cpp]: This conflict was caused by incorrect reference to windows.h header file, which has been removed already.  So we can restore its name.

File: src/bindings/c/tests/ov_test.cpp [hunk 5]
```diff
@@ -24,7 +24,7 @@ std::string plugins_xml_std = TestDataHelpers::generate_ieclass_xml_path("plugin
 #endif
 const char* plugins_xml = plugins_xml_std.c_str();
 
-std::map<ov_element_type_e, size_t> element_type_size_map = {{ov_element_type_e::BOOLEAN, 8},
+std::map<ov_element_type_e, size_t> element_type_size_map = {{ov_element_type_e::BOOLEAN, 1},
```
Comment (ilyachur) [inline src/bindings/c/tests/ov_test.cpp:27]: Why did you change size to 1?
Boolean size if I remember right is not equal to `u1` and the size is 8. Or am I wrong?

Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.cpp]: you are right, boolean is defined as 8bits data.

File: src/bindings/c/src/ov_core.cpp [hunk 6]
```diff
@@ -318,3 +321,78 @@ void ov_core_versions_free(ov_core_version_list_t* versions) {
...
+            object = core->object->compile_model(model_path, dev_name, property);
+        } else {
+            object = core->object->compile_model(model_path, property);
+        }
+        std::unique_ptr<ov_compiled_model_t> _compiled_model(new ov_compiled_model_t);
+        _compiled_model->object = std::make_shared<ov::CompiledModel>(std::move(object));
+        *compiled_model = _compiled_model.release();
+    }
+    CATCH_OV_EXCEPTIONS
+    return ov_status_e::OK;
```
Comment (ilyachur) [inline src/bindings/c/src/ov_core.cpp:396]: Why do we need to duplicate the function body? Can we convert wchar to char and re-call the base method?

File: src/bindings/c/tests/ov_test.cpp (hunk 5 above)
Comment (ilyachur) [reply src/bindings/c/tests/ov_test.cpp]: I think will be better to have tests to validate that sizes of element type is the same with C++ data types.

**Review by ilyachur — APPROVED** [review]
@riverlijunjie I think comments can be solved in the next PR.

**Comment by riverlijunjie** [comment]
> @riverlijunjie I think comments can be solved in the next PR.

OK, I will submit new PR to resolve above issue

File: src/bindings/c/src/ov_core.cpp (hunk 6 above)
Comment (riverlijunjie) [reply src/bindings/c/src/ov_core.cpp:396]: ok

File: src/bindings/c/tests/ov_test.hpp [hunk 7]
```diff
@@ -51,3 +67,58 @@ inline static std::vector<uint8_t> content_from_file(const char* filename, bool
     }
     return result;
 }
+
+#ifdef OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
+extern const std::vector<std::wstring> test_unicode_postfix_vector;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.hpp:72]: Original thought is to share it to multiple test case files, but it seem it is only used in ov_core_test.cpp, so we can move it from ov_test.hpp and ov_test.cpp, only put it into ov_core_test.cpp.

File: src/bindings/c/tests/ov_test.cpp (hunk 5 above)
Comment (riverlijunjie) [reply src/bindings/c/tests/ov_test.cpp]: ok, will provide same element type test with C++ data type.

File: src/bindings/c/src/ov_core.cpp (hunk 6 above)
Comment (riverlijunjie) [reply src/bindings/c/src/ov_core.cpp:396]: @ilyachur maybe it cannot for this function due to C va_arg() doesn't support nested calling.
